### ISSUE TITLE
extend banning functionality to cover contact in other journals/communities

### ISSUE THREAD

**Issue by rahaeli** [dreamwidth/dw-free#1661]
Currently, when you ban someone, that person cannot:

* leave a comment in your journal
* send you a PM
* buy you paid time or points in the shop
* send you points 

They can, however:

* reply to your comments in someone else's journal
* reply to your comments in a community
* reply to your entries in a community

In other words, banning applies to the journal, not to the person behind it. This has historically caused problems: a not-unknown tactic when someone has banned you is to find a comment they made in a mutual friend's journal or a comment or entry they made in an unrelated community, and reply to them there. It was originally done that way so banning someone couldn't affect their ability to comment elsewhere on the site, but it's hard to explain to people and leads to issues.

Banning should be changed to apply everywhere on the site. If Alice has banned Bob, Bob should not be able to:

* post a comment to a post Alice has made in a community
* post a reply to a comment Alice has made in a community or to another journal
* post a reply to a comment Alice has left in Bob's journal (whether it was left before the ban or after)

Some details:

* If Carol has posted an entry in a community and Alice has replied to that entry, Bob should not be able to reply to Alice's comment. Bob should still be able to post a separate top-level reply to Carol's community post.

* If Carol has posted an entry in a community, Alice has replied to that entry, and David has replied to Alice, Bob should not be able to reply to Alice, but should be able to reply to David.

* If Alice has posted a community entry, Bob should not be able to reply to that entry at all, whether it's a top-level reply or in reply to Carol or David.

* If Alice and Bob were formerly friends and Alice had left comments in Bob's journal, Bob should not be able to go back to those comments and reply to them.

I'm sure there are other use cases I'm not thinking of off the top of my head, but basically, if it's something that would fire a comment notification ("Bob replied to your comment in..."), the comment shouldn't be allowed.

The error message can be the same as it is if Bob is trying to comment in Alice's journal after being banned. (I don't know what it is off the top of my head, but something like "You cannot reply to this [comment|community entry], because Alice has banned you from replying to them" or some such is fine.)

**Comment by swaldman3** [comment]
claim (tentatively)

**Comment by swaldman3** [comment]
Note to capture a conversation that just happened at cesy's in-person hack day:

This spec, as written, means that A can ban B, then put a comment in B's journal that B cannot reply to. Some users may find obnoxious the idea that there can be comments in their journal to which they can't reply.

Discussion with @kaberett concluded that the probable reason for this is, after A has banned B, to prevent B from going through all of A's comments in B's journal and replying to them as a way of reaching A.

Probably worth explaining this (less confusingly than above) in the code tour when it's pushed.

**Comment by rahaeli** [comment]
Yes, the "Alice bans Bob, Bob goes back to find all of Alice's old comments in Bob's journal and replies to them" scenario happens a bunch. (Last reported instance, which was part of the motivation for opening this bug, had Bob repeatedly replying to Alice's old comments with things like "I miss you", etc, and when Alice deleted each replied-to comment, Bob found a different one.)

**Comment by swaldman3** [comment]
@rahaeli question: A has banned B. A posts in a community. C comments on the post. Should B be able to reply to C's comment? 
Or to put it another way, in a community, is B banned from posting *any* comments on A's posts, or just from direct replies to things that A has written?

**Comment by rahaeli** [comment]
Yes! I tried to be clear in the "some details" above, but:

Assume Alice has banned Bob.

*If Alice posts an entry in a community, Bob should not be able to reply to that entry in any way, whether top-level or in reply to someone else. 

* If Carol posts an entry in a community, and Alice replies to Carol, Bob should not be able to reply to Alice, but should be able to reply top-level (to Carol) or in reply to anyone else (other than Alice).

* If Carol posts an entry in a community, Alice replies to Carol, and David replies to Alice, Bob should not be able to reply to Alice, but should be able to reply to David.

**Comment by swaldman3** [comment]
Boh, sorry, you did explain exactly this case but it did not register in my mind! Thanks for repeating it for the hard-of-reading ;)

**Comment by rahaeli** [comment]
It's cool, it's not like I've never completely glossed over a bullet point or two in my time. ;)

My basic guiding impulse in wanting this is that it's *really common* for people to find out they're banned from someone's journal and then try to wiggle around it by continuing to contact the person. The two most common methods are creating another account to evade the ban (ToS violation), or following someone to a third-party journal/community to continue the dispute. Especially with RP, which happens mostly in communities/third-party journals, this means a lot of cases of someone getting contacted by someone they don't want to contact them anymore, and it's a fucking pain for the person who did the banning and a major driver of ToS reports.

So, generally speaking, err on the side of "no, that contact shouldn't be allowed". Pretty much any scenario where Bob is speaking directly to Alice shouldn't be possible. A good guiding question to ask: if you assume Alice has notifications of replies to her comments turned on but has not subscribed to all comments on a particular entry, if Bob's comment would fire a comment notification to Alice, that comment should not be allowed.

**Comment by swaldman3** [comment]
PR 2329 is a partial fix to this - it does the requested checks on reply, but not quickreply.

Please unassign me from this.

**Comment by zorkian** [comment]
This has been partially fixed as noted in the above PR.

**Comment by Sophira** [comment]
Would there be any recourse for community admins/staff members? I'm envisioning these scenarios:

1. Alice has banned Bob, who is the community admin for C. Alice posts or comments somewhere in C, and Bob wants to officially reply on behalf of the community but cannot because of the ban. Should Bob be able to use the "Admin Post" feature (thereby marking the comment specially) to bypass this, or is the only recourse to ban Alice from the community?
2. Alice has banned all DW staff members from their journal. It seems to me like there must be a need for DW staff members to be able to officially respond to Alice on her posts or comments, because otherwise Alice could, for example, spread a rumour about DW  which staff could not respond to.

**Comment by rahaeli** [comment]
1) Yes, Bob shouldn't be able to reply to Alice even using the "mark as admin" -- either get another admin (if there is one) to reply to her, or remove her from the community. (I really don't see a good chance of Alice wanting to stick in a community that Bob alone runs if she's banned Bob...)

2) Huh, that's an edge case I never woulda thought of even with my suspicious bastard hat on; good work on out-suspicious-bastarding me! (Not sarcasm; I'm impressed.) Yeah, official site-staff accounts should always be able to reply to anyone in official communities, even if banned from the commenter's personal journal; the recourse there for someone who doesn't want to hear from us is to avoid commenting in official communities. Bans of site staff accounts should still work in other personal journals and in non-official communities.
